bayesian/utility/v2/component_analyses.py

```diff
@@ -125,22 +125,64 @@ def get_known_unknown_pkgs(
         ecosystem: str, graph_response: Dict,
         normalised_input_pkgs: List) -> Tuple[List[Dict], Set[Package]]:
     """Analyse Known and Unknown Packages."""
+    if ecosystem == 'golang':
+        stack_recommendation, db_known_packages = get_stack_recommendation_for_golang(
+            ecosystem, graph_response, normalised_input_pkgs)
+    else:
+        stack_recommendation, db_known_packages = get_stack_recommendation(
+            ecosystem, graph_response, normalised_input_pkgs)
+
+    input_dependencies = set(normalised_input_pkgs)
+    unknown_pkgs: Set = input_dependencies.difference(db_known_packages)
+    return stack_recommendation, unknown_pkgs
+
+
+def get_stack_recommendation(ecosystem, graph_response, normalised_input_pkgs):
+    """Get Stack Recommendation for All Ecosystems, Except Golang.
+
+    Unlike Golang, No Multiple loops are required as "given_vr == version."
+    :param graph_response: Graph Response
+    :param normalised_input_pkgs: Input Packages.
+    :return: Stack Recomm. and DB_Known_Packages.
+    """
     stack_recommendation = []
     db_known_packages = set()
-    for package, input_pkg in zip(
-            graph_response.get('result', {}).get('data'), normalised_input_pkgs):
+    for package in graph_response.get('result', {}).get('data'):
         pkg_name = package.get('package').get('name', [''])[0]
         pkg_vr = package.get('version').get('version', [''])[0]
-        pkg_recomendation = CABatchResponseBuilder(ecosystem).\
-            generate_recommendation(package, input_pkg.given_version)
+        pkg_recomendation = CABatchResponseBuilder(ecosystem). \
+            generate_recommendation(package, pkg_vr)
         stack_recommendation.append(pkg_recomendation)
         known_package_flow(ecosystem, pkg_recomendation["package"], pkg_recomendation["version"])
         db_known_packages.add(normlize_packages(pkg_name, pkg_vr,
-                                                given_version=input_pkg.given_version))
+                                                given_version=pkg_vr))
+    return stack_recommendation, db_known_packages
 
-    input_dependencies = set(normalised_input_pkgs)
-    unknown_pkgs: Set = input_dependencies.difference(db_known_packages)
-    return stack_recommendation, unknown_pkgs
+
+def get_stack_recommendation_for_golang(ecosystem, graph_response, normalised_input_pkgs):
+    """Get Stack Recommendation for Golang.
+
+    In Golang, we have to loop over normalised_input_pkgs to fetch
+    "given_version" and feed that to response builder.
+    :param ecosystem: Ecosystem
+    :param graph_response: Graph Response
+    :param normalised_input_pkgs: Input Packages.
+    :return: Stack Recomm. and DB_Known_Packages.
+    """
+    stack_recommendation = []
+    db_known_packages = set()
+    for package in graph_response.get('result', {}).get('data'):
+        for input_pkg in normalised_input_pkgs:
```

Comment:
Can we change this to a map and use the same code for all ecosystems? This will also result in this loop being O(n)

Reply:
Mapping added for normalised_input_pkgs in the case of Golang.

> use the same code for all ecosystems.

IMO mapping for other ecosystems doesn't really a Gain here. https://github.com/fabric8-analytics/fabric8-analytics-server/blob/82fd64e9c7f68c280cf6f264730a5366a87a09b1/bayesian/utility/v2/component_analyses.py#L150 

PTAL

Reply:
we can get rid of duplicate code as there is not much difference in terms time.

Reply:
Ack.